konduit-serving-core/src/main/java/ai/konduit/serving/verticles/inference/InferenceVerticle.java

```diff
@@ -99,28 +105,57 @@ public void init(Vertx vertx, Context context) {
         setupWebServer();
     }
 
-    protected void setupWebServer() {
-        int portValue = inferenceConfiguration.getServingConfig().getHttpPort();
-        if (portValue == 0) {
-            String portEnvValue = System.getenv(VerticleConstants.PORT_FROM_ENV);
-            if (portEnvValue != null) {
-                portValue = Integer.parseInt(portEnvValue);
+    protected void setupWebServer() throws Throwable {
+        String portEnvValue = System.getenv(VerticleConstants.PORT_FROM_ENV);
+        if (portEnvValue != null) {
+            try {
+                port = Integer.parseInt(portEnvValue);
+            } catch (NumberFormatException e) {
+                log.error("Environment variable value isn't a valid port number",  e);
+
+                port = inferenceConfiguration.getServingConfig().getHttpPort();
             }
+        } else {
+            port = inferenceConfiguration.getServingConfig().getHttpPort();
+        }
+
+        if (port < 0 || port > 0xFFFF) {
+            port = 0; // a port of 0 will automatically be resolved into an available port.
         }
 
-        final int portValueFinal = portValue;
         List<PipelineStep> steps = inferenceConfiguration.getSteps();
         final int nSteps = steps == null ? 0 : steps.size();
-        vertx.createHttpServer()
+
+        startVertxServer(port, nSteps);
+
+        countDownLatch.await();
+
+        if(throwable != null) throw throwable;
+    }
+
+    private void startVertxServer(int port, int nSteps) {
+        new Thread(() -> vertx.createHttpServer()
                 .requestHandler(router)
                 .exceptionHandler(Throwable::printStackTrace)
-                .listen(portValueFinal, inferenceConfiguration.getServingConfig().getListenHost(), listenResult -> {
-                    if (listenResult.failed()) {
-                        log.error("Could not start HTTP server on port {}", portValueFinal);
-                        listenResult.cause().printStackTrace();
+                .listen(port, inferenceConfiguration.getServingConfig().getListenHost(), result -> {
+                    if (result.failed()) {
+                        Throwable cause = result.cause();
+
+                        if(cause instanceof BindException && cause.getMessage().equalsIgnoreCase("Address already in use")) {
+                            log.info("Port: {} is already in use. Trying another port...", port);
+                            startVertxServer(0, nSteps);
```

Comment:
Couple of things here
(a) we should probably only retry with a new port if the user wants a random port. If they want a _specific, fixed port_ (as set in their config) we should just propagate an exception instead of giving them something different
(b) Let's add some sort of "maximum number of attempts" to this? Suppose something is really wrong with the system, we don't want KS trying indefinitely if it can never succeed

Reply:
So, the maximum number of attempts didn't make sense to me after we don't want to do automatic port resolution. Now it will just throw an error if it isn't able to start the server on the first go. The port number `0` is responsible for automatic port resolution